### ISSUE TITLE
object list (GET Bucket) fails when first 1002 keys are all deleted

### ISSUE THREAD

**Issue by shino** [basho/riak_cs#654]
- With Riak 1.4.x and Riak CS 1.4.x.
- Enable `fold_objects_for_list_keys`.

Steps to reproduce:

```
$ s3cmd mb s3://test/
$ mkdir a
$ for i in {0001..1002}; do echo ${i} > a/${i}.txt; done   # in zsh
$ s3cmd put --recursive a s3://test
$ s3cmd del --recursive --force s3://test
$ s3cmd ls s3://test     # !!FAIL!!
```

Crash log

```
2013-08-26 16:37:28 =ERROR REPORT====
** State machine <0.12130.0> terminating 
** Last message in was {26551215,{done,<<"g20AAAAKYS8xMDAyLnR4dA==">>}}
** When State == waiting_object_list
**      Data  == {state,<0.271.0>,{list_objects_request_v1,<<"test">>,1000,undefined,<<"/">>,undefined}, [snip]
** Reason for termination = 
** {function_clause,[{lists,last,[[]],[{file,"lists.erl"},{line,162}]},{riak_cs_list_objects_fsm_v2,make_start_key,1,[{file,"/home/shino/work/basho/git/riak_cs-1.4/src/riak_cs_list_objects_fsm_v2.erl"},{line,359}]},{riak_cs_list_objects_fsm_v2,make_2i_request,2,[{file,"/home/shino/work/basho/git/riak_cs-1.4/src/riak_cs_list_objects_fsm_v2.erl"},{line,289}]},{riak_cs_list_objects_fsm_v2,respond,4,[{file,"/home/shino/work/basho/git/riak_cs-1.4/src/riak_cs_list_objects_fsm_v2.erl"},{line,252}]},{gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,494}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,227}]}]}
```

Direct cause:

`CommonPrefixes` is empty at https://github.com/basho/riak_cs/blob/release/1.4/src/riak_cs_list_objects_fsm_v2.erl#L356 .

**Comment by kuenishi** [comment]
This is issued via ZD ticket `#5553` - I'd be happy if included in next release (So I'm tagging as 1.5.0).

**Comment by kuenishi** [comment]
Also, without `?delimiter=/` in query url with s3curl (or else) this doesn't happen.

**Comment by reiddraper** [comment]
Working on reproducing today.

**Comment by reiddraper** [comment]
Tentative fix is:

``` patch
diff --git a/src/riak_cs_list_objects_fsm_v2.erl b/src/riak_cs_list_objects_fsm_v2.erl
index a1cf396..7742f4e 100644
--- a/src/riak_cs_list_objects_fsm_v2.erl
+++ b/src/riak_cs_list_objects_fsm_v2.erl
@@ -351,7 +351,7 @@ make_start_key(#state{object_list_ranges=[], req=Request}) ->
     make_start_key_from_marker(Request);
 make_start_key(State=#state{object_list_ranges=PrevRanges,
                             common_prefixes=CommonPrefixes}) ->
-    case last_result_is_common_prefix(State) of
+    case last_result_is_common_prefix(State) andalso CommonPrefixes =/= [] of
         true ->
             LastPrefix = lists:last(lists:sort(ordsets:to_list(CommonPrefixes))),
             skip_past_prefix_and_delimiter(LastPrefix);
```

**Comment by reiddraper** [comment]
@shino, @kuenishi  Let us know if you think this is worth cutting a 1.4.1.

**Comment by kuenishi** [comment]
I think most of 1.4.0 users might have been affected , so this is worth 1.4.1. I'd like users to experience the blazingly-fast bucket listing!

**Comment by shino** [comment]
Addressed by #655
